File: Graphs/dijkstra.py
'''
This is an implementation of Dijkstra's algorithm tuned to the specifications
of the following problem:
    http://codeforces.com/contest/20/problem/C

Python 3 is a little too slow for CF though, so this one gets a TLE.
See the C++ translation for a solution that passes.
'''

import heapq

def dijkstra(start, end, adj_list, paths):
    # Priority queue contrains triplet (cost, cur_node, prev_node).
    priority_q = [(0, start, 0)]
    while priority_q:
        cost, node, prev = heapq.heappop(priority_q)
        if paths[node] >= 0:
            # We've already reached this node.
            continue
        paths[node] = prev

        if node == end:
            break

        for weight, nbr in adj_list[node]:
            if paths[nbr] >= 0:
                continue
            heapq.heappush(priority_q, (cost + weight, nbr, node))

def main():
    n, m = map(int, input().split())

    # Adjacency list stores tuples of (weight, node).
    adj_list = [[] for i in range(n + 1)]
    for i in range(m):
        a, b, w = map(int, input().split())
        adj_list[a].append((w, b))
        adj_list[b].append((w, a))

    # Paths stores the previous node in the optimal path
    # to a given node. 0 is a dummy start node.
    paths = [-1 for i in range(n + 1)]
    dijkstra(1, n, adj_list, paths)

    # Reconstruct optimal path.
    best_path = [n]
    while paths[best_path[-1]] > 0:
        best_path.append(paths[best_path[-1]])

    if best_path[-1] == 1:
        print(*best_path[::-1])
    else:
        print(-1)

if __name__ == '__main__':
    main()

File: Graphs/dijkstra.cpp
/*
 * As close to a line by line translation of dijkstra.py as I could get.
 * This one passes all of CF's test cases.
 */
#include <algorithm>
#include <iostream>
#include <queue>
#include <tuple>
#include <vector>

using namespace std;

void dijkstra(int start, int end, const vector<tuple<int, int> > adj_list[], int paths[]) {
        // Cost, current node, previous node.
        using CNP = tuple<int, int, int>;
        priority_queue<CNP, vector<CNP>, greater<CNP> > q;
        q.push(make_tuple(0, start, 0));
        while (!q.empty()) {
                auto cnp = q.top();
                q.pop();

                auto cost = get<0>(cnp), node = get<1>(cnp);
                if (paths[node] >= 0) continue;
                paths[node] = get<2>(cnp);

                if (node == end) return;

                for (const auto& nbr : adj_list[node]) {
                        if (paths[get<1>(nbr)] >= 0) continue;
                        q.push(make_tuple(cost + get<0>(nbr), get<1>(nbr), node));
                }
        }
}

int main() {
        int N, M;
        cin >> N >> M;

        vector<tuple<int, int> > adj_list[N + 1];
        for (int i = 0; i < M; ++i) {
                int a, b, w;
                cin >> a >> b >> w;
                adj_list[a].push_back(make_tuple(w, b));
                adj_list[b].push_back(make_tuple(w, a));
        }

        int paths[N + 1];
        for (int i = 0; i <= N; ++i) paths[i] = -1;
        dijkstra(1, N, adj_list, paths);

        vector<int> optimal_path;
        int cur = N;
        while (cur > 0) {
                optimal_path.push_back(cur);
                cur = paths[cur];
        }

        if (cur != 0) {
                cout << "-1\n";
                return 0;
        }
        for (int i = optimal_path.size() - 1; i > 0; --i) cout << optimal_path[i] << " ";
        cout << N << "\n";

        return 0;
}
